### PR TITLE
[WiP] SSL certificates + letsencrypt

### PR DESCRIPTION
**Pull request by wvengen**
This implements SSL certificates with letsencrypt (#1).
On start, a dummy certificate is generated. This is necessary for HAProxy to start at all, and useful for initial testing. [acme.sh](http://acme.sh/) is used as a simple letsencrypt client. A cronjob runs to auto-renew certificates. On renewal, HAProxy reloads automatically (that was some work to figure out - somehow SIGHUP with the systemd wrapper wasn't able to restart the process).

Maybe a little explanation of how haproxy's auto-reload works. The acme and haproxy instance have a shared volume (mounted in `/acme.sh` on acme and in `/certs` on haproxy). Acme does its thing, and as part of the finish-hook it touches a file. The haproxy container has a custom run script that runs haproxy as a daemon. In the background, it watches the file touched in acme, and if that happens, a new haproxy process is started in a way that slowly takes over the connections from the old process (this is a little bit more complex, as you'll see when reading the code).

Pending:
- [ ] Log to stdout
- [ ] Instructions on how to generate a new letsencrypt certificate
- [ ] Move reloadcmd to a common place (so it's not needed when issuing a new certificate manually)
- [ ] Test letsencrypt for real (haven't tested it on a public box)
- [ ] Improve termination (`docker-compose stop haproxy`) _(nice to have)_

Oh, please look at the last commit only.
Note that there is still some discussion whether this is the best approach.

### CODE REVIEW

File: haproxy/haproxy.cfg [hunk 1]
```diff
@@ -10,20 +10,28 @@ global
...
-  bind *:443 ssl crt /certs
+  bind *:443 ssl crt /certs/certs
 
   default_backend foodsoft_web
+
   acl sharedlists_path path_beg /sharedlists
   use_backend sharedlists_web if sharedlists_path
 
   http-response set-header Strict-Transport-Security "max-age=16000000;"
+  http-response set-header X-Frame-Options "DENY"
```
Comment (paroga) [inline haproxy/haproxy.cfg]: I like this option, but I don't think it's a good idea to set it here. I'd move it into the corresponding backends (foodsoft, sharedlists, ...)

File: docker-compose.yml [hunk 2]
```diff
@@ -1,36 +1,72 @@
 version: '2'
 services:
-  foodsoft:
+
+  #
+  # Foodsoft (main)
+  #
+
+  foodsoft_web:
```
Comment (paroga) [inline docker-compose.yml:8]: i'd prefer to name the "base" containers without `_web` suffix, since we also don't provide a `command: ./proc-start web`

**Comment by paroga** [comment]
i extracted my `certbot` implementation and created PR #17

File: README.md [hunk 3]
```diff
@@ -1,14 +1,61 @@
...
+docker-compose up -d mariadb redis
+docker inspect foodcoopsnet_mariadb_1 | grep '"IPAddress"'
+# "IPAddress": "172.20.0.2",
+mysql -h 172.20.0.2 -u root -p
+```
+
+Then run the following SQL commands:
+
+```sql
+CREATE DATABASE foodsoft CHARACTER SET utf8 COLLATE utf8_bin;
```
Comment (paroga) [inline README.md:31]: We should use `COLLATE utf8mb4_unicode_520_ci`

File: docker-compose.yml (hunk 2 above)
Comment (wvengen) [reply docker-compose.yml:8]: I'd rather be too clear. Even though it looks a bit ugly to reference `foodsoft_web` in `extend`, it actually is what we do.

Comment (paroga) [reply docker-compose.yml:8]: yes, but it's also the DEFAULT, so no need to name it explicit. if you want to be clear then you also need to at the `command`stuff, which is simply not necessary

**Comment by wvengen** [comment]
Closed in favour of #17.

**Pull request closed without merging**